### PR TITLE
ec2_eip - added support for BYOIP and filtering reuse addresses by tag/value

### PR DESCRIPTION
**Pull request by tommybrecher**
##### SUMMARY

This feature adds support for allocating elastic IPs from a specific IPv4 pool (aka BYOIP).
When allocating a new EIP, provide the pool name to allocate an address from that pool. 
In addition, a mechanism for filtering reusable addresses based on a tag name / value was added.
##
#### Logic
Allocation of EIP when provided a pool:
- provide the public_ipv4_pool and the address will be alllocated from that pool
 
Filtering reusable addresses based on tags: 
  - if reuse_existing_ip_allowed is set, following logic applies
  - if tag_name is provided: only reuse addresses with that tag (regardless of it's value)
  - if tag_name is  supplemented with tag_value: ensure the tag_name (key) has a value of (tag_value)
  - if no reusable addresses -  allocate a new elastic IP from the default pool
  - if no reusable addresses and public_ipv4_pool was provided - allocate elastic IP from pool

NOTE: filtering will not take place if reuse_existing_ip_allowed is not set
##

##### ISSUE TYPE

- Feature Pull Request
##
##### COMPONENT NAME

- ec2_eip
##
##### ADDITIONAL INFORMATION

- Boto's allocate_address function does not currently support the PublicIpv4Pool attribute. 
- In order to make Amazon recognize the above attribute, a workaround was made
- The function allocate_address_from_pool() modifies the API Version before sending the request.
- It is used only for that specific use case (when attempting to allocate an EIP from a given pool).

#### Additional changes:
- Changed variable name isinstance toto is_instance (prevent overriding python's builtin function).
- Removed unused variable domain from the function ensure_absent

### CODE REVIEW

**Comment by tommybrecher** [comment]
ready_for_review

File: lib/ansible/modules/cloud/amazon/ec2_eip.py [hunk 1]
```diff
@@ -382,7 +481,10 @@ def main():
         release_on_disassociation=dict(required=False, type='bool', default=False),
         allow_reassociation=dict(type='bool', default=False),
         wait_timeout=dict(default=300, type='int'),
-        private_ip_address=dict(required=False, default=None, type='str')
+        private_ip_address=dict(required=False, default=None, type='str'),
+        tag_name=dict(required=False, default=None, type='str'),
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_eip.py]: all three of these (`required=False`, `default=None`, `type='str'`) are the defaults so can be removed

**Review by jillr — COMMENTED** [review]
Thanks @tommybrecher.  Code looks good, would like to see some integration tests.

File: lib/ansible/modules/cloud/amazon/ec2_eip.py (hunk 1 above)
Comment (tommybrecher) [reply lib/ansible/modules/cloud/amazon/ec2_eip.py]: Thanks for the feedback @willthames, should be good now I think

**Comment by tommybrecher** [comment]
> Thanks @tommybrecher. Code looks good, would like to see some integration tests.

Thanks @jillr , I was working on just that, see my last commit.

Also, something seems off with the sanity tests, they were working locally. can I re-trigger the build somehow?

**Comment by jillr** [comment]
That errors happens sometimes when shippable starts faster than the github CDN propagates - I just restarted the tests.

**Comment by jillr** [comment]
ec2_eip test passes locally for me

**Comment by tommybrecher** [comment]
What's the next step in order to get this change approved and merged?

**Review by jillr — APPROVED** [review]
@tommybrecher  Tests have all passed now, so next thing is to give @willthames a chance to re-review. Thanks for your patience!

**Comment by tommybrecher** [comment]
Hi @willthames , just in case you missed this, any chance you can have a look at the changes?

**Comment by willthames** [comment]
Merged, thanks for the nudge @tommybrecher, and for a great PR.

**Pull request merged**